index.js

```diff
@@ -54,69 +54,126 @@ class PriorityQueue {
 class PQueue {
 	constructor(opts) {
 		opts = Object.assign({
+			mustFinishDuringInterval: false,
 			concurrency: Infinity,
+			intervalCap: Infinity,
+			intervalLength: 0,
 			autoStart: true,
 			queueClass: PriorityQueue
 		}, opts);
 
 		if (!(typeof opts.concurrency === 'number' && opts.concurrency >= 1)) {
 			throw new TypeError(`Expected \`concurrency\` to be a number from 1 and up, got \`${opts.concurrency}\` (${typeof opts.concurrency})`);
 		}
+		if (!(typeof opts.intervalCap === 'number' && opts.intervalCap >= 1)) {
+			throw new TypeError(`Expected \`intervalCap\` to be a number from 1 and up, got \`${opts.intervalCap}\` (${typeof opts.intervalCap})`);
+		}
+		if (!(typeof opts.intervalLength === 'number' && Number.isFinite(opts.intervalLength) && opts.intervalLength >= 0)) {
+			throw new TypeError(`Expected \`intervalLength\` to be a finite number >= 0, got \`${opts.intervalLength}\` (${typeof opts.intervalLength})`);
+		}
+
+		this._concurrentCountIsNewIntervalCount = opts.mustFinishDuringInterval;
+		this._isPaused = opts.autoStart === false;
 
 		this.queue = new opts.queueClass(); // eslint-disable-line new-cap
 		this._queueClass = opts.queueClass;
-		this._pendingCount = 0;
-		this._concurrency = opts.concurrency;
-		this._isPaused = opts.autoStart === false;
+
+		this._concurrentCount = 0;
+		this._concurrentCap = opts.concurrency;
+
+		this._isIntervalIgnored = opts.intervalCap === Infinity || opts.intervalLength === 0;
+		this._intervalCount = 0;
+		this._intervalCap = opts.intervalCap;
+		this._intervalTime = opts.intervalLength;
```

Comment:
I didn't call it interval because I wanted to differentiate it from the other interval variables but I agree with you, the different names are confusing.